### ISSUE TITLE
Add a DB migration to migrate module settings from `analytics` to `analytics-4`

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#8082]
## Feature Description

#7923 added settings that exist in the `analytics` module and are still relevant after the removal of UA to the `analytics-4` module. It also introduced a fallback mechanism so that when SK queries these new `analytics-4` module settings but they are unavailable, their values from `analytics` are retrieved and returned.

However, during a recent audit of the implementation, I discovered that this fallback mechanism only works for SK installations that already had the Analytics module set up before #7923 was merged. If Analytics is newly set up, the default values are saved for the `analytics-4` module settings, thus not having the ability to fall back to `analytics` counterparts.

With this in mind, a DB migration should be introduced that should perform a one-time migration of the moved-over setting values. This migration should only happen as part of the same release #7932 (the issue that finally replaces everything `analytics` with `analytics-4) goes out.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A database migration should be added targeting the Site Kit release that goes out with [#7932](https://github.com/google/site-kit-wp/issues/7932).
* The database migration should do the following:
  * It should migrate values of `analytics` module settings that were copied over to `analytics-4` as part of [#7923](https://github.com/google/site-kit-wp/issues/7923).
    * The `adsenseLinked` setting should be excluded as it has been removed.
  * It should update the `googlesitekit_active_modules` option so that if the array in this option includes `analytics`, it should also add `analytics-4`.
  * It should update the `googlesitekit_dashboard_sharing` option so that if the array in this option includes sharing settings for `analytics` but not for `analytics-4`, it should add `analytics-4` and duplicate the sharing settings from `analytics` to `analytics-4`.
* The mechanism introduced in [#7923](https://github.com/google/site-kit-wp/issues/7923) within `analytics-4` to fallback to module setting values of `analytics` should be removed.

## Implementation Brief

* [x] Add `includes/Core/Util/Migration_1_120_0` class
  * Use `includes/Core/Util/Migration_1_8_1.php`, or `includes/Core/Util/Migration_1_3_0.php` as a starting point 
  * In `migrate` method invoke a method, say `migrate_legacy_settings`
  * Add `migrate_legacy_settings` method. Pull the options from `googlesitekit_analytics_settings` (Note that `analytics` module is going to be removed soon, so access the settings directly instead of using the constant `Analytics::OPTION`) and update the following keys to the `analytics-4` counterpart:
    * `accountID`
    * `adsConversionID`
    * `canUseSnippet` 
    * `trackingDisabled`
  * After successful migration update `googlesitekit_db_version` option like in the other existing migration classes.
* [x] Update `includes/Plugin::register`
  * Instantiate  `Migration_1_120_0` class and invoke `register` method after existing migration classes https://github.com/google/site-kit-wp/blob/85a20a57732e469fec0786280462435e5ccf47a9/includes/Plugin.php#L211-L212

### Test Coverage

* Add tests for `Migration_1_120_0` class

### Test 1: Before you switch to the feature branch:

* Set up SK and the Analytics module.
* Go to Analytics settings.
* Modify some options for `analytics` data store by pasting the snippets in the console, for example:
  * `googlesitekit.api.set( 'modules', 'analytics', 'settings', { adsConversionID: '12345', trackingDisabled: []} )`. This will trigger the request, you can check in `network` tab to verify that settings are updated.
  * Paste `googlesitekit.data.select('modules/analytics-4').getSettings()` to verify that `analytics-4` settings are still on their default.
* Go to the database, and update `googlesitekit_db_version` option in `wp_options` table to `1.8.1` (if it is not set to that already).
* **Now, switch to the feature branch.**
* Go back to the Analytics settings, and refresh the page - it should reflect the migration change - `Ads Conversion ID` and `Excluded from Analytics` option should say `Analytics is currently enabled for all visitors`
* Paste `googlesitekit.data.select('modules/analytics-4').getSettings()`
  * It should have migrated values for mentioned settings
* Check again the `googlesitekit_db_version` value in the database, it should have updated to the `1.123.0`

### Test 2: Before you switch to the feature branch:

* Set up SK and the Analytics module.
* Go to the database, and update `googlesitekit_db_version` option in `wp_options` table to `1.8.1`
* Now, switch to the feature branch.
* Verify that the Analytics module is still connected and functioning normally.
* Check again the `googlesitekit_db_version` value in the database, it should have updated to the `1.123.0`

### Test 3: Before you switch to the feature branch:

* Switch to SK version 1.98.0 or earlier.
* Set up SK and the Analytics module.
* Using dashboard sharing, share the Analytics module.
* Go to the database, and update `googlesitekit_db_version` option in `wp_options` table to `1.8.1`
* Now, switch to the feature branch.
* Verify that the Analytics module is still shared as it was before.
* Check again the `googlesitekit_db_version` value in the database, it should have updated to the `1.123.0`

## Changelog entry

* Migrate settings from the legacy to the new Analytics module.

**Comment by eugene-manuilov** [comment]
AC ✔️

**Comment by eugene-manuilov** [comment]
> * Add `Google\Site_Kit\Modules\Analytics_4\Migrate_Legacy_Analytics_Settings` class

@zutigrm, we use a bit different approach for migrations. See the following class as an example. We need to create one more that will migrate Analytics settings:

https://github.com/google/site-kit-wp/blob/develop/includes/Core/Util/Migration_1_8_1.php

**Comment by bethanylang** [comment]
@zutigrm Just a heads up that this issue is now unblocked so would be great to get the IB updates taken care of here as soon as possible so we can get started. Thanks!

**Comment by zutigrm** [comment]
@eugene-manuilov Thanks for the feedback, IB updated

**Comment by eugene-manuilov** [comment]
Thanks, @zutigrm. IB ✔️

**Comment by nfmohit** [comment]
Just a heads-up that I've updated the AC and IB to exclude migrating the `adsenseLinked` module setting, as it is being removed and replaced as part of #8047.

CC: @eugene-manuilov @zutigrm

**Comment by nfmohit** [comment]
Handing this over to @jimmymadon to hold this as I am soon going on some time off. Thank you Jimmy!

Just to recap, merging the PR for this issue includes a few considerations:

1. [#8082](https://github.com/google/site-kit-wp/issues/8082) implements a database migration targeted to a specific SK release. It should target (and be merged into) the same release where [#7932](https://github.com/google/site-kit-wp/issues/7932) goes out. [The PR](https://github.com/google/site-kit-wp/pull/8123/) currently [tentatively targets 1.121.0](https://github.com/google/site-kit-wp/pull/8123/files#diff-2bdee4eb80e7de4ef1223c75a26860a8e6b4c47b8d0abf4c4066d99ea0c54a8aR28), but that will most likely need to change based on the final SAM launch plans. The owner of this issue needs to be aware of the final SAM launch plans and get the target version updated to align with the same release where [#7932](https://github.com/google/site-kit-wp/issues/7932) goes out.
2. I have proposed a slightly different launch plan for SAM [here](https://10up.slack.com/archives/CBKKQEBR9/p1707554237266289), where we perform E2E testing on a feature branch before we merge [#7932](https://github.com/google/site-kit-wp/issues/7932) into `develop`. If it is agreed to go with this plan, it is necessary to merge [the PR](https://github.com/google/site-kit-wp/pull/8123/) for [#8082](https://github.com/google/site-kit-wp/issues/8082) into the said feature branch instead of `develop` for the purpose of E2E testing. Once the E2E testing is complete and approved, this feature branch can be merged into `develop`.

**Comment by nfmohit** [comment]
> * It should update the `googlesitekit_active_modules` option so that if the array in this option includes `analytics`, it should also add `analytics-4`.
> * It should update the `googlesitekit_dashboard_sharing` option so that if the array in this option includes sharing settings for `analytics` but not for `analytics-4`, it should add `analytics-4` and duplicate the sharing settings from `analytics` to `analytics-4`.

Update: I have added the above ACs as a reflection of our findings [here](https://github.com/google/site-kit-wp/issues/7932#issuecomment-1980240996). CC: @aaemnnosttv

**Comment by nfmohit** [comment]
This issue is back in code review with the new AC additions mentioned above addressed.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

- Tested on dev branch and Feature branch.
- Verified all Test 1, Test 2 and Test 3 scenarios mentioned under QAB.

@nfmohit As per last step mentioned under Test 1 db should get update.

`Paste googlesitekit.data.select('modules/analytics-4').getSettings()
It should have migrated values for mentioned settings
Check again the googlesitekit_db_version value in the database, it should have updated to the 1.123.0`


But I noticed that after pasting snippet `googlesitekit.data.select('modules/analytics-4').getSettings()` analytics settings got migrate but under settings analytics still showed AdConversion ID - 12345 and under `db googlesitekit_db_version` not updated to the 1.123.0

<details>

Last step in step 1 Db not updated and settings not updated

![image](https://github.com/google/site-kit-wp/assets/94359491/90d9a266-6e8b-4652-a7eb-79aa612767f0)

![image](https://github.com/google/site-kit-wp/assets/94359491/9edada06-b44c-4162-bd60-4fe3362ece5e)

![image](https://github.com/google/site-kit-wp/assets/94359491/dd64de08-e495-48fa-a89b-4d26b5d80dae)

</details>

**PASS CASES**

<details>

Test 1


![image](https://github.com/google/site-kit-wp/assets/94359491/0cd7c70d-59be-4bb3-9d6c-9082aec2e644)

![image](https://github.com/google/site-kit-wp/assets/94359491/ce5a5aa6-b0e7-4d63-a756-a32532b6e6fe)

![image](https://github.com/google/site-kit-wp/assets/94359491/67acef12-cd99-4329-b67e-946d4e59b557)

Before Refresh :

![image](https://github.com/google/site-kit-wp/assets/94359491/b151dd2f-a1ae-4e29-a91d-ed4660d42dc3)

![image](https://github.com/google/site-kit-wp/assets/94359491/81ae791e-e82a-4898-bc9c-f8e314f1113d)


After Refresh:

![image](https://github.com/google/site-kit-wp/assets/94359491/310a87ec-7b6d-4920-af05-ce98d41d42e4)


Test 3:


Uploading Recording #817.mp4…

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Thanks @nfmohit for updating QAB after slack discussion.
- Tested on dev branch and Feature branch.
- Verified all Test 1, Test 2 and Test 3 scenarios mentioned under QAB.
- Verifed that Test 1 cases is executed successfully and getting expected results.

<details>

Test 1:

![image](https://github.com/google/site-kit-wp/assets/94359491/af8bff0f-5484-4997-b0e7-8a5cae187117)

![image](https://github.com/google/site-kit-wp/assets/94359491/556eaf8f-1191-447f-83b6-347109e1bb19)

![image](https://github.com/google/site-kit-wp/assets/94359491/add73b07-63d8-4eda-a7e8-2d8aebcb763a)


https://github.com/google/site-kit-wp/assets/94359491/e654734d-5b27-432d-a75a-6fe6e3436d2a


Test 3:


https://github.com/google/site-kit-wp/assets/94359491/880b62f8-d0d1-47aa-ba86-16c8e261fde3

</details>

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@mohitwp @nfmohit  I ran through this ticket to ensure the SAM migration passed. For test 3, I set up Site Kit and Analytics on version `1.98.0` but could not select a property or web data stream when setting up Analytics. I ended up creating a new property and web data stream.  When I went to Analytics settings, they were blank, and I could not select Analytics from the dashboard sharing console.  At this stage, I haven't switched to the feature branch. 

Maybe it's just my test site, but I was wondering how this was tested.  

https://github.com/google/site-kit-wp/assets/73545194/9a4d802c-76c7-44bd-b960-a0edbfbef02f
